compiler/rustc_parse/src/parser/token_type.rs

```diff
@@ -0,0 +1,616 @@
+        }
+    };
+    (@binop, $op:ident) => {
+        $crate::parser::token_type::ExpTokenPair {
+            tok: &rustc_ast::token::BinOp(rustc_ast::token::BinOpToken::$op),
+            token_type: $crate::parser::token_type::TokenType::$op,
+        }
+    };
+    (@open, $delim:ident, $token_type:ident) => {
+        $crate::parser::token_type::ExpTokenPair {
+            tok: &rustc_ast::token::OpenDelim(rustc_ast::token::Delimiter::$delim),
+            token_type: $crate::parser::token_type::TokenType::$token_type,
+        }
+    };
+    (@close, $delim:ident, $token_type:ident) => {
+        $crate::parser::token_type::ExpTokenPair {
+            tok: &rustc_ast::token::CloseDelim(rustc_ast::token::Delimiter::$delim),
+            token_type: $crate::parser::token_type::TokenType::$token_type,
+        }
+    };
+
+    // `ExpKeywordPair` helper rules.
+    (@kw, $kw:ident, $token_type:ident) => {
+        $crate::parser::token_type::ExpKeywordPair {
+            kw: rustc_span::symbol::kw::$kw,
+            token_type: $crate::parser::token_type::TokenType::$token_type,
+        }
+    };
+    (@sym, $kw:ident, $token_type:ident) => {
+        $crate::parser::token_type::ExpKeywordPair {
+            kw: rustc_span::symbol::sym::$kw,
+            token_type: $crate::parser::token_type::TokenType::$token_type,
+        }
+    };
+
+    (Eq)             => { exp!(@tok, Eq,        Eq) };
+    (Lt)             => { exp!(@tok, Lt,        Lt) };
+    (Le)             => { exp!(@tok, Le,        Le) };
+    (EqEq)           => { exp!(@tok, EqEq,      EqEq) };
+    (Gt)             => { exp!(@tok, Gt,        Gt) };
+    (AndAnd)         => { exp!(@tok, AndAnd,    AndAnd) };
+    (OrOr)           => { exp!(@tok, OrOr,      OrOr) };
+    (Not)            => { exp!(@tok, Not,       Not) };
+    (Tilde)          => { exp!(@tok, Tilde,     Tilde) };
+    (At)             => { exp!(@tok, At,        At) };
+    (Dot)            => { exp!(@tok, Dot,       Dot) };
+    (DotDot)         => { exp!(@tok, DotDot,    DotDot) };
+    (DotDotDot)      => { exp!(@tok, DotDotDot, DotDotDot) };
+    (DotDotEq)       => { exp!(@tok, DotDotEq,  DotDotEq) };
+    (Comma)          => { exp!(@tok, Comma,     Comma) };
+    (Semi)           => { exp!(@tok, Semi,      Semi) };
+    (Colon)          => { exp!(@tok, Colon,     Colon) };
+    (PathSep)        => { exp!(@tok, PathSep,   PathSep) };
+    (RArrow)         => { exp!(@tok, RArrow,    RArrow) };
+    (FatArrow)       => { exp!(@tok, FatArrow,  FatArrow) };
+    (Pound)          => { exp!(@tok, Pound,     Pound) };
+    (Question)       => { exp!(@tok, Question,  Question) };
+    (Eof)            => { exp!(@tok, Eof,       Eof) };
+
+    (Plus)           => { exp!(@binop, Plus) };
+    (Minus)          => { exp!(@binop, Minus) };
+    (Star)           => { exp!(@binop, Star) };
+    (And)            => { exp!(@binop, And) };
+    (Or)             => { exp!(@binop, Or) };
```

Comment:
I'm half tempted at making this support things like `exp!(+)` and `exp!(?)` 😈

Reply:
I was initially going to do that. At one point I had `exp!(_)` instead of `exp!(Underscore)` and I was using actual keywords like `exp!(if)` instead of `exp!(If)`. A lot of tokens can be made to work but all the delimiters absolutely cannot, so I decided to use `StudlyCaps` names for all of them :)